src/components/building/building.js

```diff
@@ -0,0 +1,115 @@
+      const currentBuilding = currentUserBuildings.find(currentBuilding => currentBuilding.buildingId === buildingId);
+      const nextBuildingLevel = currentBuilding.level + 1;
+      const progressBuildingTime = buildingBuildTime * 1000;
+
+      const startBuildingTime = Date.now();
+      const endBuildingTime = startBuildingTime + progressBuildingTime;
+
+      addBuildingToBuild(0, buildingId, startBuildingTime, endBuildingTime, progressBuildingTime, nextBuildingLevel);
+
+      const buildingType = gameBuildingsTypes.find(buildingType => buildingType.id === buildingId);
+
+      buildingType.buildMaterials.forEach(material => {
+        if (material.resourceType !== 'energy') {
+          removeResources(
+            material.resourceType,
+            buildingPrice(nextBuildingLevel, buildingType.id, material.resourceType)
+          );
+        }
+      });
+    }
+  }
+
+  return gameBuildingsTypes.map(gameBuildingType => {
+    let currentUserBuilding = currentUserBuildings.find(
+      userBuilding => userBuilding.buildingId === gameBuildingType.id
+    );
+
+    let inProgressId = 0;
+    let buttonTexts = 'Build';
+
+    if (currentUserBuildingInProgress.length > 0) {
+      inProgressId = currentUserBuildingInProgress[0].buildingId;
+      buttonTexts = inProgressId === gameBuildingType.id ? 'in Progress' : buttonTexts;
+    }
+
+    buttonTexts = currentUserBuilding.level > 0 && buttonTexts !== 'in Progress' ? 'Upgrade' : buttonTexts;
+
+    const nextLevel = currentUserBuilding.level + 1;
+
+    let buildPrice = [];
+    gameBuildingType.buildMaterials.forEach(material => {
+      if (material.resourceType === 'money' || material.resourceType === 'iron') {
+        buildPrice.push(buildingPrice(nextLevel, gameBuildingType.id, material.resourceType));
+      }
+    });
+    const buildTime = buildingTime(
+      buildPrice[0] !== undefined ? buildPrice[0] : 0,
+      buildPrice[1] !== undefined ? buildPrice[1] : 0
+    );
+
+    const buildInProgressButtonDisable = currentUserBuildingInProgress.length === 0 ? false : true;
+    const buildInProgressTime =
+      currentUserBuildingInProgress.length === 0
+        ? ''
+        : timeBuilder(Math.round(currentUserBuildingInProgress[0].diffTime / 1000));
+
+    let buildPriceTextColor = [];
+    let notEnoughRescourceButtonDisable = false;
+
+    gameBuildingType.buildMaterials.forEach(buildMaterial => {
+      for (const key in currentUserResources) {
+        if (key === buildMaterial.resourceType) {
+          const currentResource = currentUserResources[key];
+          const neededResource = buildingPrice(nextLevel, gameBuildingType.id, key);
+          const enoughRescource = currentResource >= neededResource;
+          const textColor = enoughRescource ? 'black' : 'red';
+          if (!enoughRescource) {
+            notEnoughRescourceButtonDisable = true;
+          }
+          buildPriceTextColor.push(textColor);
+        }
+      }
+    });
```

Comment:
It is a bit uncommon to have all of this functionality within the return statement. I would recommend to extract this functionality to change your data and then use it later in this return statement.

Reply:
this is a separate user story where the functions are outsourced. :D

Reply:
okay 👍